bls377/g1.go

```diff
@@ -496,6 +620,86 @@ func (p *G1Jac) ScalarMulByGen(curve *Curve, scalar fr.Element) *G1Jac {
 	return p.pippenger(curve, []G1Jac{curve.g1Gen}, []fr.Element{scalar}, sGen, bGen, curve.tGenG1[:], computeT)
 }
 
+// WindowedMultiExp set p = scalars[0]*points[0] + ... + scalars[n]*points[n]
+// assume: scalars in non-Montgomery form!
+// assume: len(points)==len(scalars)>0, len(scalars[i]) equal for all i
+// algorithm: a special case of Pippenger described by Bootle:
+// https://jbootle.github.io/Misc/pippenger.pdf
+// uses all availables runtime.NumCPU()
+func (p *G1Jac) WindowedMultiExp(curve *Curve, points []G1Jac, scalars []fr.Element) *G1Jac {
+	var lock sync.Mutex
+	parallel.Execute(0, len(points), func(start, end int) {
+		var t G1Jac
+		t.multiExp(curve, points[start:end], scalars[start:end])
+		lock.Lock()
+		p.Add(curve, &t)
+		lock.Unlock()
+	}, false)
+	return p
+}
+
+// multiExp set p = scalars[0]*points[0] + ... + scalars[n]*points[n]
+// assume: scalars in non-Montgomery form!
+// assume: len(points)==len(scalars)>0, len(scalars[i]) equal for all i
+// algorithm: a special case of Pippenger described by Bootle:
+// https://jbootle.github.io/Misc/pippenger.pdf
+func (p *G1Jac) multiExp(curve *Curve, points []G1Jac, scalars []fr.Element) *G1Jac {
+	const s = 4 // s from Bootle, we choose s divisible by scalar bit length
+	const b = s // b from Bootle, we choose b equal to s
+	// WARNING! This code breaks if you switch to b!=s
+	// Because we chose b=s, each set S_i from Bootle is simply the set of points[i]^{2^j} for each j in [0:s]
+	// This choice allows for simpler code
+	// If you want to use b!=s then the S_i from Bootle are different
+	const TSize = (1 << b) - 1 // TSize is size of T_i sets from Bootle, equal to 2^b - 1
+	// Store only one set T_i at a time---don't store them all!
+	var T [TSize]G1Jac // a set T_i from Bootle, the set of g^j for j in [1:2^b] for some choice of g
+	computeT := func(T []G1Jac, t0 *G1Jac) {
+		T[0].Set(t0)
+		for j := 1; j < (1<<b)-1; j = j + 2 {
+			T[j].Set(&T[j/2]).Double()
+			T[j+1].Set(&T[(j+1)/2]).Add(curve, &T[j/2])
+		}
+	}
+	return p.pippenger(curve, points, scalars, s, b, T[:], computeT)
```

Comment:
Not part of this PR, but oh man we really need to clean up this part of multiexp.  That `computeT` parameter is ugly. 😝

Reply:
Yep, we should do that in a separate branch I think...